### ISSUE TITLE
YouTube subscription imports

### ISSUE THREAD

**Issue by stl3** [futo-org/grayjay-android#412]
After logging in, trying to import subscriptions yields a few issues:
- at the subscription window, if you accidentally click on a YouTube channel while scrolling, you have to start loading subscriptions back from the beginning which can be a tedious process if you have a large amount of subscriptions.
- for some reason it stops letting me import beyond reaching a certain channel, refusing to load more beyond that when I know there are more channels (I can parse the channel list if needed). For example in my case it doesn't go beyond when it reaches the channel "Sub Pop". Is there a limit on how many subscribed channels we can import or is this possibly a parsing issue? I do have channels with names in Japanese kanji and other languages but they appear only after the Z in alphabet, at least when it's displayed on YouTube Android when I scroll through my subscribed channels list. 
Let me know how I can help, there doesn't seem to be errors showing, maybe there's a log file in grayjay but I haven't looked yet.

PS - is there a more sane automated way to import subscriptions? For example let it import automatically without needing to manually click "Load more" which tends to become cumbersome. Let it respect the rate limiting, but give it a random cooldown between 5-10 seconds before fetching the next amount. App also needs to be able to continue this in background when screen shuts off because it restarts back from the beginning when that happens.
PPS - sorry but had to also mention there needs to be a way to separate subscription videos from mixing with what's on YouTube homepage videos. Sometimes I just want to see the new videos from subscriptions.

**Comment by Kelvin-FUTO** [comment]
When you click import we already know all the channels you are subscribed to. The issue is getting the metadata for each individual channel.
Seeing how import is a one-time thing, the current phased approach was an easy way to prevent people from getting errors. 

The total import we've tested with Youtube I believe is about 300 channels.
There should be no theoretical limit on the amount of subscriptions (assuming <1500 or so practically), the experience however might be sub-part right now with having more than 150-200 subscriptions, which we are improving.
Japanese characters should not pose an issue.
You have not said how many channels you have so its hard to really make out much.

Im a bit confused about your home page comment. Your subscriptions are a different tab. If you hate the home tab you can disable it in settings (under Manage Tabs). 
The Home tab represents the home pages of your respective plugins, if you're logged in this may contain items from your subscriptions. You can also disable the home page for individual plugins in their respective detail settings.

**Comment by stl3** [comment]
> When you click import we already know all the channels you are subscribed to. The issue is getting the metadata for each individual channel. Seeing how import is a one-time thing, the current phased approach was an easy way to prevent people from getting errors.
> 
> The total import we've tested with Youtube I believe is about 300 channels. There should be no theoretical limit on the amount of subscriptions (assuming <1500 or so practically), the experience however might be sub-part right now with having more than 150-200 subscriptions, which we are improving. Japanese characters should not pose an issue. You have not said how many channels you have so its hard to really make out much.
> 
It was about 1k+ subscriptions (under 1.5k). I've removed some older ones which no longer upload so right now it's just shy of 1k.

> Im a bit confused about your home page comment. Your subscriptions are a different tab. If you hate the home tab you can disable it in settings (under Manage Tabs). The Home tab represents the home pages of your respective plugins, if you're logged in this may contain items from your subscriptions. You can also disable the home page for individual plugins in their respective detail settings.

Sorry I should have clarified a bit. On the Home tab (the screen that shows when the app is loaded), I can see my subscriptions (videos) mixed with videos that I assume are also from the Home area in youtube - everytime I pull down to refresh it shows slightly different videos from there (or somewhere) mixed in with what looks like channels I've subscribed to, making it a little confusing. I've only enabled the Youtube plugin for now so I run into less problems to easier iron them out.

This is my [manage tabs settings](https://imgur.com/kjGWyhM)
This is [what home shows](https://imgur.com/tfOz5Y4)
Am I meant to be seeing an additional tab called "Subscriptions" next to home?

Note: I've been using arm64-v7a

**Comment by xephos1one** [comment]
> I just saw someone elses picture of how it's meant to look like and I'm definitely missing the "Subscriptions" tab even though it's enabled.

Have you tried to disable the Subscriptions Tab and re-enable it afterwards?

**Comment by stl3** [comment]
> > I just saw someone elses picture of how it's meant to look like and I'm definitely missing the "Subscriptions" tab even though it's enabled.
> 
> Have you tried to disable the Subscriptions Tab and re-enable it afterwards?

Yes, when I first noticed that it was missing, I went to manage tabs and enabled and disabled, restarted app - still not showing. I tried seeing if rearranging it would help, still the same. I cleared cache, cleared data, uninstalled, reinstalled - still missing. I tried v7a/v8a (did not try universal) but still not showing. Should I try universal or is there something else I can check?

**Comment by Kelvin-FUTO** [comment]
Right, so your subscriptions tab is missing all-together. We hide the subscriptions tab if you have no subscriptions (those under creators). If you see no profiles in the Creators tab, it means you have no subscriptions, and thus the tab is hidden.
Keep in mind that simply logging into Youtube does not mean your subscriptions are automatically the same. You can import your Youtube subscriptions be clicking "Import Subscriptions" in the Youtube plugin. (In the future we might add an auto-sync)

Maybe something went wrong there. We are going to disable this mechanism anyway soon and point people to subscribing to creators, which should avoid this in the future.
Maybe try changing the order of the tabs, perhaps fixes it temporarily?

To clarify, we do NOT mix subscriptions into home page. If you see videos there that are part of your subscriptions that just means Youtube is returning those videos from their home page (especially if logged in, this might be a close match). By Youtube behavior, these videos will also disappear after you watch them if you have playback tracking enabled for Youtube.

**Comment by stl3** [comment]
I was able to get the subscriptions tab working but on another phone. However, the importing of subscriptions issue was still there (not getting the full list) and I found out a couple of channels giving issues:
https://www.youtube.com/channel/UCOpNcN46UbXVtpKMrmU4Abg - Gaming (https://www.youtube.com/@gaming)
https://www.youtube.com/channel/UC4R8DWoMoI7CAwX8_LjQHig - Live (https://www.youtube.com/@live)
These were youtube curated channels for the Gaming and Live sections and not user channels. In my youtube subscriptions following the Gaming and Live from youtube puts it as a subscribed channel (though going to the @ for the channels don't let you really unsubscribe from them - I can only remove it through searching which then shows an unsubscribe pulldown from the search page oddly)
![image](https://github.com/futo-org/grayjay-android/assets/32008952/460512f2-cd01-447e-9de5-faa02b527790)
Cannot unsubscribe from channel unless through search
![image](https://github.com/futo-org/grayjay-android/assets/32008952/6680e8b8-a527-4f66-9531-af84662af51c)
![image](https://github.com/futo-org/grayjay-android/assets/32008952/840541df-cd51-4376-b256-045f6655b530)

Initially it wasn't showing any errors in Grayjay until I removed more channels, which it then showed "Failed to fetch" for 2 channels which were the gaming and live channels.
After having purged quite a few older channels and the 2 channels mentioned above, I was able to get more or less all the remaining channels imported. Basically I had to remove a channel from my youtube subscriptions, check to see any changes in the import window in grayjay, and re-add the channel back if it wasn't the one causing the problem and rinse/repeat until it started showing the proper list of subscriptions I had (very long process if you have over 1k subscriptions so anyone who is going to try this, be warned, it is very time consuming).

**Comment by Kelvin-FUTO** [comment]
I suspect the issue is that those channels are not actually channels but more like collections, Grayjay expects a channel but finds such a weird page, and Grayjay fails on those. I'll try to investigate that soon. You obviously shouldn't have to deal with that.

**Comment by Zvonimir-FUTO** [comment]
This seems to be fixed in latest version (264) of Grayjay. Feel free to open a new issue if you are still encountering this.
